src/routes/transactions/mappers/common/twap-order.mapper.ts

```diff
@@ -195,37 +193,110 @@ export class TwapOrderMapper {
-    const completeStatuses = [
-      OrderStatus.Fulfilled,
-      OrderStatus.Cancelled,
-      OrderStatus.Expired,
-    ];
+    const now = new Date();
+    const activePart = args.twapParts.find((part) => {
+      return part.validTo > Math.floor(now.getTime() / 1_000);
+    });
 
-    for (let i = 0; i < orders.length; i++) {
-      const order = orders[i];
+    return activePart ?? null;
+  }
 
-      // Return the status of the last part
-      if (i === orders.length - 1) {
-        return order.status;
+  private async getPartOrders(args: {
+    partsToFetch: Array<GPv2OrderParameters>;
+    chainId: string;
+    safeAddress: `0x${string}`;
+  }): Promise<Array<KnownOrder> | null> {
+    const orders: Array<KnownOrder> = [];
+
+    for (const part of args.partsToFetch) {
+      const partFullAppData = await this.swapsRepository.getFullAppData(
+        args.chainId,
+        part.appData,
+      );
+
+      if (!this.swapAppsHelper.isAppAllowed(partFullAppData)) {
+        throw new Error(
+          `Unsupported App: ${partFullAppData.fullAppData?.appCode}`,
+        );
+      }
+
+      const orderUid = this.gpv2OrderHelper.computeOrderUid({
+        chainId: args.chainId,
+        owner: args.safeAddress,
+        order: part,
+      });
+      const order = await this.swapsRepository
+        .getOrder(args.chainId, orderUid)
+        .catch(() => {
+          this.loggingService.warn(
+            `Error getting orderUid ${orderUid} from SwapsRepository`,
+          );
+        });
+
+      if (!order || order.kind == OrderKind.Unknown) {
+        // Without every order it's not possible to determine executed amounts/fees
+        return null;
       }
 
-      // If the part is complete, continue to the next part
-      if (completeStatuses.includes(order.status)) {
-        continue;
+      if (!this.swapAppsHelper.isAppAllowed(order)) {
+        throw new Error(`Unsupported App: ${order.fullAppData?.appCode}`);
       }
 
-      return order.status;
+      orders.push(order as KnownOrder);
+    }
+
+    return orders;
+  }
+
+  private async getOrderStatus(args: {
+    chainId: string;
+    safeAddress: `0x${string}`;
+    twapParts: Array<GPv2OrderParameters>;
+    partOrders: Array<KnownOrder> | null;
+    activeOrderUid: `0x${string}` | null;
+    executionDate: Date | null;
+  }): Promise<OrderStatus> {
+    if (!args.executionDate) {
+      return OrderStatus.PreSignaturePending;
+    }
+
+    const orderUid =
+      args.activeOrderUid ??
+      this.gpv2OrderHelper.computeOrderUid({
+        chainId: args.chainId,
+        owner: args.safeAddress,
+        // If there is no "active" part, we use the last one to get the status
+        order: args.twapParts.slice(-1)[0],
+      });
+
+    // If we already have the order, there is no need to fetch it again
+    if (args.partOrders) {
+      const order = args.partOrders.find((order) => order.uid === orderUid);
+      if (order) {
+        return order.status;
```

Comment:
Maybe I'm misunderstanding this, but if you just return the order status, wouldn't that be wrong for "fulfilled" orders? If the status is fulfilled for one part, it doesn't mean that this applies to the whole TWAP.

Reply:
The logic regarding status generation is different as of https://github.com/safe-global/safe-client-gateway/pull/1785#discussion_r1691083471 which should now solve this.